examples/fluids/navierstokes.c

```diff
@@ -1235,23 +1266,79 @@ int main(int argc, char **argv) {
     ierr = DMGetLocalVector(dm, &Qexactloc); CHKERRQ(ierr);
     ierr = VecGetSize(Qexactloc, &lnodes); CHKERRQ(ierr);
 
-    ierr = ICs_PetscMultiplicity(op_ics, xcorners, q0ceed, dm, Qexactloc, Qexact,
-                                 restrictq, &ctxSetup, ftime); CHKERRQ(ierr);
+    ierr = ICs_FixMultiplicity(op_ics, xcorners, q0ceed, dm, Qexactloc, Qexact,
+                               restrictq, &ctxSetup, ftime); CHKERRQ(ierr);
 
     ierr = VecAXPY(Q, -1.0, Qexact);  CHKERRQ(ierr);
     ierr = VecNorm(Q, NORM_MAX, &norm); CHKERRQ(ierr);
     CeedVectorDestroy(&q0ceed);
     ierr = PetscPrintf(PETSC_COMM_WORLD,
                        "Max Error: %g\n",
                        (double)norm); CHKERRQ(ierr);
+    // Clean up vectors
+    ierr = DMRestoreLocalVector(dm, &Qexactloc); CHKERRQ(ierr);
+    ierr = VecDestroy(&Qexact); CHKERRQ(ierr);
   }
 
   // Output Statistics
   ierr = TSGetStepNumber(ts,&steps); CHKERRQ(ierr);
   if (!test) {
     ierr = PetscPrintf(PETSC_COMM_WORLD,
                        "Time integrator took %D time steps to reach final time %g\n",
-                       steps,(double)ftime); CHKERRQ(ierr);
+                       steps, (double)ftime); CHKERRQ(ierr);
+  }
+  // Output numerical values from command line
+  ierr = VecViewFromOptions(Q, NULL, "-vec_view"); CHKERRQ(ierr);
+
+  if (test) { // compare reference solution values with current run
+    PetscViewer viewer;
+    char filepath[PETSC_MAX_PATH_LEN];
+    PetscReal testtol;
+    // Create reference solution file name and assign relative error tolerance
+    if (!implicit) { // Explicit test run
+      testtol = 1E-5;
+      ierr = PetscStrncpy(filepath,
+                          "examples/fluids/tests-output/fluids-navierstokes-explicit.bin",
+                          sizeof filepath);
+      CHKERRQ(ierr);
+    }
+    else { // Implicit test runs
+      testtol = 5E-4;
+      if (stab != STAB_SUPG) {
+        ierr = PetscStrncpy(filepath,
+                            "examples/fluids/tests-output/fluids-navierstokes-implicit-nostab.bin",
+                            sizeof filepath);
+        CHKERRQ(ierr);
+      }
+      else {
+        ierr = PetscStrncpy(filepath,
+                            "examples/fluids/tests-output/fluids-navierstokes-implicit-supgstab.bin",
+                            sizeof filepath);
+        CHKERRQ(ierr);
+      }
+    }
```

Comment:
Maybe we should make a testType enum, testTypes strings, testData struct, and testOptions global variable. I found myself working with similar if chains when I rewrote the PETSc BPs. These additions would remove the if chains with repeated code and make it easier to add future tests.

Reply:
Done. Please let me know if it looks good to you

Reply:
I had a couple of comments, but overall I think this is much cleaner. Thanks.